src/nfa/nfa.cc

```diff
@@ -1354,3 +1562,501 @@ void Mata::Nfa::ShortestWordsMap::update_current_words(LengthWordsPair& act, con
+            product.make_final(intersect_state_to);
+        }
+    }
+    else
+    {
+        intersect_state_to = product_map[intersect_state_pair_to];
+    }
+    intersect_transitions.states_to.insert(intersect_state_to);
+}
+
+void Intersection::compute_for_same_symbols(const TransSymbolStates& lhs_state_transitions,
+                                            const TransSymbolStates& rhs_state_transitions)
+{
+    // Create transition from the pair_to_process to all pairs between states to which first transition goes and states
+    // to which second one goes.
+    TransSymbolStates intersection_transition(lhs_state_transitions.symbol);
+    for (State this_state_to: lhs_state_transitions.states_to)
+    {
+        for (State other_state_to: rhs_state_transitions.states_to)
+        {
+            create_product_state_and_trans(this_state_to, other_state_to, intersection_transition);
+        }
+    }
+
+    add_product_transition(intersection_transition);
+}
+
+void Mata::Nfa::Intersection::initialize_pairs_to_process()
+{
+    for (State this_initial_state : lhs.initialstates)
+    {
+        for (State other_initial_state : rhs.initialstates)
+        {
+            handle_initial_state_pairs(this_initial_state, other_initial_state);
+        }
+    }
+}
+
+void Mata::Nfa::Intersection::handle_initial_state_pairs(State lhs_initial_state, State rhs_initial_state)
+{
+    StatePair this_and_other_initial_state_pair(lhs_initial_state, rhs_initial_state);
+    State new_intersection_state = product.add_new_state();
+
+    product_map[this_and_other_initial_state_pair] = new_intersection_state;
+    pairs_to_process.insert(this_and_other_initial_state_pair);
+
+    product.make_initial(new_intersection_state);
+    if (lhs.has_final(lhs_initial_state) && rhs.has_final(rhs_initial_state))
+    {
+        product.make_final(new_intersection_state);
+    }
+}
+
+void Mata::Nfa::Intersection::add_product_transition(const TransSymbolStates& intersection_transition)
+{
+    auto& intersect_state_transitions{ product.transitionrelation[product_map[pair_to_process]] };
+    auto symbol_transitions{ intersect_state_transitions.find(intersection_transition) };
+    if (symbol_transitions == intersect_state_transitions.end())
+    {
+        intersect_state_transitions.push_back(intersection_transition);
+    }
+    else if (symbol_transitions->states_to != intersection_transition.states_to)
+    {
+        TransSymbolStates new_symbol_transitions{ intersection_transition.symbol,
+                                                  symbol_transitions->states_to.Union(intersection_transition.states_to) };
+        intersect_state_transitions.remove(*symbol_transitions);
+        intersect_state_transitions.push_back(new_symbol_transitions);
+    }
+}
+
+Nfa Mata::Nfa::concatenate(const Nfa& lhs, const Nfa& rhs)
+{
+     return Concatenate(lhs, rhs).get_result();
+}
+
+void Concatenate::compute_concatenation()
+{
+    map_states_to_result_states();
+
```

Comment:
I think it's OK, to not have empty line between statements. It will not impact readability :)

Reply:
I can see nothing wrong with removing the empty line. The empty line is a remnant of the first implementation of the function, where there were some more complicated operations and there were clear sections for mapping of states to result states and the following code.

I will refactor.